### PR TITLE
Home work 23

### PR DESCRIPTION
**Pull request by ExQA**

### CODE REVIEW

File: src/test/java/org/glovo/service/OrderServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        products.add(product3);
+
+        ArrayList<Product> products2 = new ArrayList<>();
+        products2.add(product1);
+        products2.add(product2);
+
+        GlovoOrder existingOrder = new GlovoOrder(1L, products, 36.12, 3, null, null);
+        GlovoOrder updatedOrder = new GlovoOrder(1L, products2, 24.12, 2, null, null);
+
+        // Set up the mock behavior
```
Comment (mkharkhuta) [inline src/test/java/org/glovo/service/OrderServiceTest.java]: Why do you need it here?

File: src/test/java/org/glovo/service/OrderServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        when(orderRepository.getReferenceById(1L)).thenReturn(order);
+        when(orderRepository.save(order)).thenReturn(order);
+
+        GlovoOrder result = orderService.removeProduct(1, 1);
+
+        assertEquals(order, result);
+        verify(orderRepository, times(1)).getReferenceById(1L);
+        verify(orderRepository, times(1)).save(order);
+    }
+}
```
Comment (mkharkhuta) [inline src/test/java/org/glovo/service/OrderServiceTest.java:163]: There are no negative test, edge cases

File: src/test/java/org/glovo/service/OrderServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    @Mock
+    private ProductService productService;
+
+    @InjectMocks
+    private OrderService orderService = new OrderService();
+    @Captor
+    private ArgumentCaptor<GlovoOrder> orderCaptor;
+
+    @Test
+    public void testGetAllOrders() {
```
Comment (mkharkhuta) [inline src/test/java/org/glovo/service/OrderServiceTest.java:53]: Please use provide more complex test name, to understend what is tested, whith what condition and what the expected resut

File: src/test/java/org/glovo/service/OrderServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+        assertEquals(expectedOrders, actualOrders);
+    }
+
+    @Test
+    void getByIdTest() {
+        Product product1 = new Product(1, "apple", 12.12, 2);
+        Product product2 = new Product(2, "apple", 12, 2);
+        Product product3 = new Product(3, "apple", 12, 2);
+
```
Comment (mkharkhuta) [inline src/test/java/org/glovo/service/OrderServiceTest.java]: As this is a duplicate from previous, it can be extracted to a separate method

File: src/test/java/org/glovo/service/OrderServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    @Mock
+    private ProductRepository productRepository;
+
+    @Mock
+    private OrderRepository orderRepository;
+
+    @Mock
+    private ProductService productService;
+
+    @InjectMocks
```
Comment (mkharkhuta) [inline src/test/java/org/glovo/service/OrderServiceTest.java:47]: Why do you need this annotation?

File: src/test/java/org/glovo/service/ProductServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        Product actualProduct = productService.create(product);
+        assertEquals(expectedProduct, actualProduct);
+    }
+
+    @Test
+    public void testDeleteProduct() {
+        productService.delete(1);
+
+        verify(productRepository, times(1)).deleteById(1L);
+    }
```
Comment (mkharkhuta) [inline src/test/java/org/glovo/service/ProductServiceTest.java:72]: The same comments

**Review by mkharkhuta — COMMENTED** [review]
Good, but please, check the comments

**Comment by mkharkhuta** [comment]
Also, please, add only relevant changes to the PR

File: src/main/java/org/glovo/service/OrderService.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Date;
+import java.util.List;
+
+
+@Service
+public class OrderService {
+    @Autowired
+    private OrderRepository orderRepository;
+    @Autowired
+    private ProductService productService;
```
Comment (IvashchenkoIvanIgorevich) [inline src/main/java/org/glovo/service/OrderService.java:19]: IMO better to use @AllArgsConstructor

File: src/main/java/org/glovo/service/OrderService.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import org.glovo.entity.Product;
+import org.glovo.repository.OrderRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+
```
Comment (IvashchenkoIvanIgorevich) [inline src/main/java/org/glovo/service/OrderService.java:13]: blank rows should be deleted

File: src/test/java/org/glovo/service/OrderServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    @Mock
+    private OrderRepository orderRepository;
+
+    @Mock
+    private ProductService productService;
+
+    @InjectMocks
+    private OrderService orderService = new OrderService();
```
Comment (IvashchenkoIvanIgorevich) [inline src/test/java/org/glovo/service/OrderServiceTest.java:48]: We don't need create instance, if you use @InjectMocks it means that Spring create and get instance from context.

**Review by IvashchenkoIvanIgorevich — APPROVED** [review]
